src/Std/Time/Notation.lean

```diff
@@ -0,0 +1,246 @@
+  | .H p => do `(Std.Time.Modifier.H $(← convertNumber p))
+  | .m p => do `(Std.Time.Modifier.m $(← convertNumber p))
+  | .s p => do `(Std.Time.Modifier.s $(← convertNumber p))
+  | .S p => do `(Std.Time.Modifier.S $(← convertFraction p))
+  | .A p => do `(Std.Time.Modifier.A $(← convertNumber p))
+  | .n p => do `(Std.Time.Modifier.n $(← convertNumber p))
+  | .N p => do `(Std.Time.Modifier.N $(← convertNumber p))
+  | .V => `(Std.Time.Modifier.V)
+  | .z p => do `(Std.Time.Modifier.z $(← convertZoneName p))
+  | .O p => do `(Std.Time.Modifier.O $(← convertOffsetO p))
+  | .X p => do `(Std.Time.Modifier.X $(← convertOffsetX p))
+  | .x p => do `(Std.Time.Modifier.x $(← convertOffsetX p))
+  | .Z p => do `(Std.Time.Modifier.Z $(← convertOffsetZ p))
+
+private def convertFormatPart : FormatPart → MacroM (TSyntax `term)
+  | .string s => `(.string $(Syntax.mkStrLit s))
+  | .modifier mod => do `(.modifier $(← convertModifier mod))
+
+private def syntaxNat (n : Nat) : MacroM (TSyntax `term) := do
+  let info ← MonadRef.mkInfoFromRefPos
+  pure { raw := Syntax.node1 info `num (Lean.Syntax.atom info (toString n)) }
+
+private def syntaxString (n : String) : MacroM (TSyntax `term) := do
+  let info ← MonadRef.mkInfoFromRefPos
+  pure { raw := Syntax.node1 info `str (Lean.Syntax.atom info (toString n)) }
+
+private def syntaxInt (n : Int) : MacroM (TSyntax `term) := do
+  match n with
+  | .ofNat n => `(Int.ofNat $(Syntax.mkNumLit <| toString n))
+  | .negSucc n => `(Int.negSucc $(Syntax.mkNumLit <| toString n))
+
+private def syntaxBounded (n : Int) : MacroM (TSyntax `term) := do
+ `(Std.Time.Internal.Bounded.LE.ofNatWrapping $(← syntaxInt n) (by decide))
+
+private def syntaxVal (n : Int) : MacroM (TSyntax `term) := do
+ `(Std.Time.Internal.UnitVal.ofInt $(← syntaxInt n))
+
+private def convertOffset (offset : Std.Time.TimeZone.Offset) : MacroM (TSyntax `term) := do
+ `(Std.Time.TimeZone.Offset.ofSeconds $(← syntaxVal offset.second.val))
+
+private def convertTimezone (tz : Std.Time.TimeZone) : MacroM (TSyntax `term) := do
+ `(Std.Time.TimeZone.mk $(← convertOffset tz.offset) $(Syntax.mkStrLit tz.name) $(Syntax.mkStrLit tz.abbreviation) false)
+
+private def convertPlainDate (d : Std.Time.PlainDate) : MacroM (TSyntax `term) := do
+ `(Std.Time.PlainDate.ofYearMonthDayClip $(← syntaxInt d.year) $(← syntaxBounded d.month.val) $(← syntaxBounded d.day.val))
+
+private def convertPlainTime (d : Std.Time.PlainTime) : MacroM (TSyntax `term) := do
+ `(Std.Time.PlainTime.mk $(← syntaxBounded d.hour.val) $(← syntaxBounded d.minute.val) ⟨true, $(← syntaxBounded d.second.snd.val)⟩ $(← syntaxBounded d.nanosecond.val))
+
+private def convertPlainDateTime (d : Std.Time.PlainDateTime) : MacroM (TSyntax `term) := do
+ `(Std.Time.PlainDateTime.mk $(← convertPlainDate d.date) $(← convertPlainTime d.time))
+
+private def convertZonedDateTime (d : Std.Time.ZonedDateTime) (identifier := false) : MacroM (TSyntax `term) := do
```

Comment:
Thinking one more time about this, I think all this is maybe really `Lean.ToExpr` instead? Maybe @kmill has some thoughts here.